### DISCUSSION TITLE
"No-op" merges?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2649] [Q&A]
Often in Git, I want to ensure my feature branch is up-to-date with the trunk, but not make any changes if it already incorporates the tip of the trunk. I generally do this with a merge (though I realize that rebase is also a common strategy).

In `jj`, it appears that `merge` always creates a new commit (and somewhat weirdly, it permits the parent commit to be empty, which I expect is a pretty typical case when merging something with `@`). Is there a way to create a merge commit if and only if none of the parents contain every other parent? I.e. for two parents, the `xor` of the commit sets isn't empty; for more than two, every parent is "behind" at least one other parent in the Git sense.

**Comment by martinvonz** [comment]
IIUC what you want, `jj new 'all:heads(heads(::@ ~ empty()) | main)'` probably does that. What that does:

1. `heads(::@ ~ empty())` gives the first non-empty commit(s) that are ancestors of `@`. Typically that's just one commit, unless you have some merge commits.
2. `heads(heads(::@ ~ empty()) | main)` gives only the newest commit between the commits from step 1 and `main`.
3. `all:` allows you to specify all those commits (typically one or two) as new parents.

Btw, `merge` is basically an alias for `new`.

**Reply by BatmanAoD** [reply]
Interesting; thank you! Any chance of making that...easier?

**Reply by BatmanAoD** [reply]
...also, why does `heads(heads(::@ ~ empty()) | main)` include `main` IFF `main` has a commit that isn't on `@`? I see that this works, I just don't quite understand the revset logic here.

`heads(::@ ~ empty())` seems useful as some kind of alias anyway; it's almost always what I want when I use `branch set`, but currently I just use `-r @-` and, if I forget the `-`, I need to `--allow-backwards` to fix it.

**Reply by BatmanAoD** [reply]
...actually, for `branch set`, if `heads(::@ ~ empty())` is _multiple_ commits (i.e. `@` is a merge), then I actually want `@`, since obviously `branch set` can't take multiple commits. I'm sure there's some way to express this in revset language too, but an alias that just picks "the current latest non-empty commit" would be _really_ helpful for updating branches.

**Reply by BatmanAoD** [reply]
In fact... could "the latest non-empty commit" be made the _default_ target of `branch set`? The current default of `@` is usually wrong, in the sense that branches (as far as I can tell, and in my own usage at least) are _primarily_ useful for pushing changes to a remote, and by default `jj` won't push commits without a description, so having a branch on a descriptionless commit is usually just annoying.

**Reply by martinvonz** [reply]
I suppose you can define a revset alias like this:

```toml
revset-aliases.here = "heads(::@ ~ (empty() ~ merges()))"
```

Then you can do e.g. `jj branch set my-branch -r here` or `jj new 'all:here|main'`. If it turns out to be common for people to define such an alias, we can make it a built-in.

Regarding the `all:` bit and the desire to not create a merge if one parent is an ancestor of another, #2600 is related to that. As I said on that bug, I think we should actually be better at *not* dropping redundant ancestors. Perhaps the right solution here is to add a `jj git sync` command (#1039) and give it a config option to choose between rebase and merge. If configured to merge, it would presumably behave the way you're asking for `jj merge` to behave. Does that sound good to you?

**Reply by BatmanAoD** [reply]
Yeah, `sync` is semantically the operation I want and makes sense here, I think; and it would indeed be helpful if the `merge` variant of `sync` behaved this way, especially since it's close to how Git behaves.

**Reply by BatmanAoD** [reply]
Odd, this worked for me earlier today in one repo, but in another, I'm getting this error:

```
❯ : jj new 'all(heads(heads(::@ ~ empty()) | master))'
Error: Failed to parse revset:  --> 1:5
  |
1 | all(heads(heads(::@ ~ empty()) | master))
  |     ^----------------------------------^
  |
  = Invalid arguments to revset function "all": Expected 0 arguments
Hint: Prefix the expression with 'all' to allow any number of revisions (i.e. 'all:all(heads(heads(::@ ~ empty()) | master))').


❯ : jj new 'all:all(heads(heads(::@ ~ empty()) | master))'
Error: Failed to parse revset:  --> 1:5
  |
1 | all(heads(heads(::@ ~ empty()) | master))
  |     ^----------------------------------^
  |
  = Invalid arguments to revset function "all": Expected 0 arguments
```

**Reply by martinvonz** [reply]
I think you meant `jj new 'all:heads(heads(::@ ~ empty()) | master)'`. `all:` is a special prefix, not part of the revset syntax. However, there's also a revset function called `all()`. That function doesn't take any arguments.

**Reply by BatmanAoD** [reply]
Oh, you're right; I thought I'd copied the command the same way both times, but I must have done it with `all:heads(...)` the first time.
